File: app/src/main/java/com/grohden/niceanimals/modules/NetModule.java
package com.grohden.niceanimals.modules;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.grohden.niceanimals.shibe.service.ShibeService;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
public class NetModule {

    String mBaseUrl;

    public NetModule(String baseUrl) {
        this.mBaseUrl = baseUrl;
    }

    @Provides
    @Singleton
    Retrofit provideRetrofit(Gson gson) {
        return new Retrofit
                .Builder()
                .baseUrl(mBaseUrl)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
    }

    @Provides
    @Singleton
    Gson provideGson() {
        GsonBuilder gsonBuilder = new GsonBuilder();
        return gsonBuilder.create();
    }

    @Provides
    @Singleton
    public ShibeService provideShibeService(Retrofit retrofit) {
        return retrofit.create(ShibeService.class);
    }
}

File: app/src/main/java/com/grohden/niceanimals/modules/NetModule.kt
package com.grohden.niceanimals.modules

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.grohden.niceanimals.shibe.service.ShibeService
import dagger.Module
import dagger.Provides
import retrofit2.Retrofit
import retrofit2.adapter.java8.Java8CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Singleton

@Module
class NetModule(internal var mBaseUrl: String) {

    @Provides
    @Singleton
    internal fun provideRetrofit(gson: Gson): Retrofit {
        return Retrofit.Builder()
                .baseUrl(mBaseUrl)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(Java8CallAdapterFactory.create())
                .build()
    }

    @Provides
    @Singleton
    internal fun provideGson(): Gson {
        val gsonBuilder = GsonBuilder()
        return gsonBuilder.create()
    }

    @Provides
    @Singleton
    fun provideShibeService(retrofit: Retrofit): ShibeService {
        return retrofit.create(ShibeService::class.java)
    }
}
